icelake/src/io_v2/functional_writer/precompute_partition_writer.rs

```diff
@@ -0,0 +1,305 @@
+        } else {
+            return Err(Error::new(
+                ErrorKind::IcebergDataInvalid,
+                "Expect partition type to be struct",
+            ));
+        };
+        if !is_same {
+            return Err(Error::new(
+                ErrorKind::ArrowError,
+                format!(
+                    "Expect partition type {:?}, but got {:?}",
+                    expect_partition_type,
+                    remove_field.data_type()
+                ),
+            ));
+        }
+
+        let arrow_partition_type_fields =
+            if let DataType::Struct(fields) = self.partition_type.clone().try_into()? {
+                fields
+            } else {
+                unreachable!()
+            };
+        let row_converter = RowConverter::new(vec![SortField::new(DataType::Struct(
+            arrow_partition_type_fields.clone(),
+        ))])
+        .map_err(|e| Error::new(ErrorKind::ArrowError, e.to_string()))?;
+
+        Ok(PrecomputedPartitionedWriter {
+            inner_writers: HashMap::new(),
+            inner_buidler: self.inner,
+            schema: new_schema_builder.finish().into(),
+            partition_type: self.partition_type,
+            row_converter,
+            partition_col_idx: self.partition_col_idx,
+        })
+    }
+}
+
+/// Partition append only writer
+pub struct PrecomputedPartitionedWriter<B: IcebergWriterBuilder> {
+    inner_writers: HashMap<OwnedRow, (B::R, StructValue)>,
+    partition_type: Any,
+    row_converter: RowConverter,
+    inner_buidler: B,
+    schema: SchemaRef,
+    partition_col_idx: usize,
+}
+
+#[async_trait::async_trait]
+impl<B: IcebergWriterBuilder> IcebergWriter for PrecomputedPartitionedWriter<B> {
+    type R = <<B as IcebergWriterBuilder>::R as IcebergWriter>::R;
+
+    /// Write a record batch. The `DataFileWriter` will create a new file when the current row num is greater than `target_file_row_num`.
+    async fn write(&mut self, mut batch: RecordBatch) -> Result<()> {
+        // Remove the partition value from batch.
+        let partition_value = batch.remove_column(self.partition_col_idx);
+        let value_array = struct_to_anyvalue_array_with_type(
+            partition_value
+                .as_any()
+                .downcast_ref::<StructArray>()
+                .unwrap(),
+            self.partition_type.clone(),
+        )?;
+
+        // Group the batch by row value.
+        // # TODO
+        // Maybe optimize the alloc and clone
+        let rows = self
+            .row_converter
+            .convert_columns(&[partition_value.clone()])
+            .map_err(|e| Error::new(ErrorKind::ArrowError, e.to_string()))?;
+        let mut filters = HashMap::new();
+        rows.into_iter().enumerate().for_each(|(row_id, row)| {
+            filters
+                .entry(row.owned())
+                .or_insert_with(|| (vec![false; batch.num_rows()], row_id))
+                .0[row_id] = true;
+        });
```

Comment:
This is kind of inefficient, how about generating an array of `(partition_value, row_id)` and then sort them?

Reply:
Good point!